lib/src/cache/cache.dart

```diff
@@ -0,0 +1,241 @@
+import 'dart:async';
+
+import 'package:meta/meta.dart';
+
+import 'package:w_common/disposable.dart';
+import 'package:w_common/func.dart';
+
+/// Immutable payload that indicates a change in a [Cache].
+class CacheContext<TIdentifier, TValue> {
+  /// Identifies a single instance of a [TValue]
+  final TIdentifier id;
+
+  /// The current value stored in the [Cache] for the attributed [TIdentifier].
+  final TValue value;
+
+  CacheContext(this.id, this.value);
+}
+
+/// Maintains a reference to a given [TValue] associated with a [TIdentifier].
+///
+/// References are retained for the lifecycle of the instance of the [Cache],
+/// unless explicitly removed.
+class Cache<TIdentifier, TValue> extends Object with Disposable {
+  /// The backing store for values in the [Cache].
+  Map<TIdentifier, Future<TValue>> _cache = <TIdentifier, Future<TValue>>{};
+
+  // ignore: close_sinks
+  StreamController<CacheContext<TIdentifier, TValue>> _didRemoveController =
+      new StreamController<CacheContext<TIdentifier, TValue>>.broadcast();
+
+  // ignore: close_sinks
+  StreamController<CacheContext<TIdentifier, TValue>> _didUpdateController =
+      new StreamController<CacheContext<TIdentifier, TValue>>.broadcast();
+
+  /// A collection of pending removals for a given [TIdentifier].
+  Map<TIdentifier, Completer<Null>> _removalCompleters =
+      <TIdentifier, Completer<Null>>{};
+
+  Cache() {
+    [_didRemoveController, _didUpdateController]
+        .forEach(manageStreamController);
+  }
+
+  /// The stream of events that indicate a removal [Cache].
+  Stream<CacheContext<TIdentifier, TValue>> get didRemove =>
+      _didRemoveController.stream;
+
+  /// The stream of events that indicate a change to a cached [TValue].
+  Stream<CacheContext<TIdentifier, TValue>> get didUpdate =>
+      _didUpdateController.stream;
+
+  /// Returns a value from the cache for a given [TIdentifier]. If the
+  /// [TIdentifier] is not present in the cache the value returned by the given
+  /// [valueFactory] is added to the cache and returned.
+  ///
+  /// All calls to [get] await a call to the [onGet] lifecycle method before
```

Comment:
Just to get this straight in my head, if I call `get` twice in a row with the same identifier without awaiting the resulting future, will `onGet` be called once or twice?